DAOs enabledModAware

kennyqiyenkan opened the pull request:
Added the following:
- EnabledAwareDAO
  - both .js and .java
- LastModifiedDAO
  - both .js and .java

kennyqiyenkan committed:
EnabledAwareDAO/LastModifiedAwareDAO

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master'

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into DAOs-enabledModAware

kennyqiyenkan committed:
Merge branch 'master' into DAOs-enabledModAware

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into DAOs-enabledModAware

kennyqiyenkan commented:
I signed it!

kgrgreer commented on src/foam/nanos/auth/LastModifiedAwareDAO.java:
```diff
@@ -0,0 +1,20 @@
+import foam.core.*;
+import java.time.LocalDateTime;
+
+public abstract class LastModifiedAwareDAO
```
Shouldn't be abstract.

kgrgreer commented on src/foam/nanos/auth/LastModifiedAwareDAO.js:
```diff
@@ -0,0 +1,22 @@
+    {
+      name: 'put',
+      code: function(value) {
+        value.lastModified = new Date().now();
```
lastModified is a Date, so remove the .now().

kgrgreer commented on tools/classes.js:
```diff
@@ -48,6 +48,7 @@ var classes = [
   'foam.mlang.sink.Count',
   'foam.nanos.boot.NSpec',
   'foam.nanos.auth.EnabledAware',
+  'foam.nanos.auth.EnabledAwareDAO',
```
Remove from here but put in src/foam/files.js.

kgrgreer commented:
Now add the Java decorators, which are the primary goal.

adamvy commented on src/foam/nanos/auth/LastModifiedAwareDAO.js:
```diff
@@ -0,0 +1,22 @@
+foam.CLASS({
+  package: 'foam.nanos.auth',
+  name: 'LastModifiedAwareDAO',
+  extends: 'foam.dao.ProxyDAO',
```
I think these would be better served as DAODecorators rather than ProxyDAOs.

kgrgreer replied:
DAODecorator makes modifying the data on the way out easier, but these decorators are only adding behaviour on put(), so I don't think it buys anything here.

adamvy commented on src/foam/nanos/auth/LastModifiedAwareDAO.java:
```diff
@@ -0,0 +1,20 @@
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.auth;
```
Are these classes really nanos specific? Or would they do better in a more generate purpose package?

kgrgreer replied:
LastModifiedAware and EnabledAware aren't but LastModifiedByAware is, and it seemed like they should be in the same directory. I thought about putting them in foam.pattern, but you generally think of patterns as being more complex. Perhaps just putting them in foam.dao would be better, and then add a constant to LastModifiedByAware for the PRINCIPAL_KEY, so that it's not directly tied to the auth system.

kennyqiyenkan committed:
Added Java Version of EnabledAwareDAO

kennyqiyenkan committed:
Misc Changes Following Comments

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into DAOs-enabledModAware

# Conflicts:
#	src/files.js

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into DAOs-enabledModAware

kgrgreer commented:
@kennyqiyenkan Please move this code  to the foam.dao package instead of foam.nanos.auth. Thx

kennyqiyenkan committed:
Merge remote-tracking branch 'foam-framework/master' into DAOs-enabledModAware

kennyqiyenkan committed:
Changed EnabledAware

kennyqiyenkan committed:
Removed redundant DAO implement

kennyqiyenkan committed:
Relocated to appropriate packages

and changed the packages appropriately.

kennyqiyenkan committed:
Removed DAO implement

kennyqiyenkan committed:
Updated files.js to reflect relocations

kennyqiyenkan committed:
Used SUPER instead of super as `super` is reserved

kgrgreer commented on java_src/foam/dao/EnabledAwareDAO.java:
```diff
@@ -0,0 +1,26 @@
+
+public class EnabledAwareDAO
+  extends ProxyDAO
+  implements EnabledAware
```
Doesn't need to implement EnabledAware. Only the objects it's filtering need to.

kgrgreer commented on java_src/foam/dao/EnabledAwareDAO.java:
```diff
@@ -0,0 +1,26 @@
+  implements EnabledAware
+{
+  public Sink select(Sink s, Integer skip, Integer limit, Comparator order, Predicate predicate) {
+    return super.select(s, skip, limit, order, EQ(this.EnabledAware.ENABLED, true));
```
this could be a constant so that you don't need to create a new one for each call

kgrgreer commented on java_src/foam/dao/LastModifiedAwareDAO.java:
```diff
@@ -0,0 +1,23 @@
+{
+  public FObject put(FObject value) {
+    ((LastModifiedAware)value).lastModified = LocalDateTime.now();
+    super.put(value);
```
The delegate could return a different object, so the correct semantics would be:
return super.put(value);

kgrgreer commented on src/foam/dao/LastModifiedAwareDAO.js:
```diff
@@ -0,0 +1,21 @@
+      name: 'put',
+      code: function(value) {
+        value.lastModified = new Date();
+        SUPER.put(value);
```
return SUPER(value);

kennyqiyenkan committed:
Modified along comments

- removed unnecessary implements
- return SUPER in js file
- return the delegate instead of value
- made a Constant variable for Predicate

kgrgreer commented on java_src/foam/dao/EnabledAwareDAO.java:
```diff
@@ -14,13 +14,13 @@
   extends ProxyDAO
-  implements EnabledAware
 {
+  private Predicate PREDICATE_ = EQ(this.EnabledAware.ENABLED, true);
```
Should be 'final' and 'static' to be a constant. Otherwise you're still creating multiple instants and users could update the value.

kennyqiyenkan committed:
Actually made it constant
